File: app/src/main/java/org/xdty/callerinfo/model/CallRecord.java
package org.xdty.callerinfo.model;

public class CallRecord {

    private long ring = -1;
    private long hook = -1;
    private long idle = -1;
    private long ringDuration = -1;
    private long callDuration = -1;

    private String mLogNumber;
    private String mLogName;
    private String mLogGeo;

    public String getLogName() {
        return mLogName;
    }

    public void setLogName(String name) {
        this.mLogName = name;
    }

    public void setLogName(String name, boolean append) {
        if (append) {
            appendName(name);
        } else {
            setLogName(name);
        }
    }

    public String getLogGeo() {
        return mLogGeo;
    }

    public void setLogGeo(String name) {
        this.mLogGeo = name;
    }

    public void ring() {
        ring = System.currentTimeMillis();
    }

    public void hook() {
        hook = System.currentTimeMillis();
    }

    public long getHook() {
        return hook;
    }

    public void idle() {
        idle = System.currentTimeMillis();

        if (isIncoming()) {
            if (hook == -1) { // missed or hangup incoming call
                ringDuration = idle - ring;
                callDuration = 0;
            } else { // answered incoming call
                ringDuration = hook - ring;
                callDuration = idle - hook;
            }
        } else {  // outgoing call
            ringDuration = 0;
            callDuration = idle - hook;
        }
    }

    public boolean isIncoming() {
        return ring != -1;
    }

    public long ringDuration() {
        return ringDuration;
    }

    public long callDuration() {
        return callDuration;
    }

    public void reset() {
        ring = -1;
        hook = -1;
        idle = -1;
        ringDuration = -1;
        callDuration = -1;

        mLogNumber = null;
        mLogGeo = null;
        mLogName = null;
    }

    public long time() {
        return ring != -1 ? ring : hook;
    }

    public String getLogNumber() {
        return mLogNumber;
    }

    public void setLogNumber(String mLogNumber) {
        this.mLogNumber = mLogNumber;
    }

    public void appendName(String s) {
        if (mLogName == null || mLogName.isEmpty()) {
            mLogName = "";
        }
        mLogName += " " + s;
    }

    public boolean isValid() {
        return mLogNumber != null && !mLogNumber.isEmpty();
    }

    public boolean isNameValid() {
        return mLogName != null && !mLogName.isEmpty();
    }

    public boolean isGeoValid() {
        return mLogGeo != null && !mLogGeo.isEmpty();
    }

    public boolean matchName(String keyword) {
        return mLogName != null && mLogName.contains(keyword);
    }

    public boolean matchGeo(String keyword) {
        return mLogGeo.contains(keyword);
    }

    public boolean matchNumber(String keyword) {
        return mLogNumber != null && mLogNumber.startsWith(keyword);
    }

    public boolean isActive() {
        return ring != -1 || hook != -1 || idle != -1;
    }

    public boolean isAnswered() {
        return isIncoming() && callDuration > 0;
    }

    public boolean isEqual(String number) {
        return (mLogNumber == null && number == null)
                || (mLogNumber != null && number != null && mLogNumber.equals(number));
    }
}

File: app/src/main/java/org/xdty/callerinfo/model/CallRecord.kt
package org.xdty.callerinfo.model

class CallRecord {
    private var ring: Long = -1
    var hook: Long = -1
        private set
    private var idle: Long = -1
    private var ringDuration: Long = -1
    private var callDuration: Long = -1
    var logNumber: String = ""
    var logName: String = ""
    var logGeo: String = ""

    val isIncoming: Boolean
        get() = ring != -1L

    val isValid: Boolean
        get() = logNumber.isNotEmpty()

    val isNameValid: Boolean
        get() = logName.isNotEmpty()

    val isGeoValid: Boolean
        get() = logGeo.isNotEmpty()

    val isActive: Boolean
        get() = ring != -1L || hook != -1L || idle != -1L

    val isAnswered: Boolean
        get() = isIncoming && callDuration > 0

    fun setLogName(name: String, append: Boolean) {
        if (append) {
            appendName(name)
        } else {
            logName = name
        }
    }

    fun ring() {
        ring = System.currentTimeMillis()
    }

    fun hook() {
        hook = System.currentTimeMillis()
    }

    fun idle() {
        idle = System.currentTimeMillis()
        if (isIncoming) {
            if (hook == -1L) { // missed or hangup incoming call
                ringDuration = idle - ring
                callDuration = 0
            } else { // answered incoming call
                ringDuration = hook - ring
                callDuration = idle - hook
            }
        } else { // outgoing call
            ringDuration = 0
            callDuration = idle - hook
        }
    }

    fun ringDuration(): Long {
        return ringDuration
    }

    fun callDuration(): Long {
        return callDuration
    }

    fun reset() {
        ring = -1
        hook = -1
        idle = -1
        ringDuration = -1
        callDuration = -1
        logNumber = ""
        logGeo = ""
        logName = ""
    }

    fun time(): Long {
        return if (ring != -1L) ring else hook
    }

    private fun appendName(s: String) {
        logName += " $s"
    }

    fun matchName(keyword: String?): Boolean {
        return logName.contains(keyword!!)
    }

    fun matchGeo(keyword: String?): Boolean {
        return logGeo.contains(keyword!!)
    }

    fun matchNumber(keyword: String?): Boolean {
        return logNumber.startsWith(keyword!!)
    }

    fun isEqual(number: String?): Boolean {
        return (number == null || logNumber == number)
    }
}